Common/src/geometry/CPhysicalGeometry.cpp

```diff
@@ -5218,57 +5182,70 @@ void CPhysicalGeometry::SetPositive_ZArea(CConfig *config) {
 
-  for (iElem = 0; iElem < nElem; iElem++)
-
-  /*--- Loop over all the nodes of an element ---*/
-
-    for (iNode = 0; iNode < elem[iElem]->GetnNodes(); iNode++) {
-      iPoint = elem[iElem]->GetNode(iNode);
-
-      /*--- Store the element into the point ---*/
+    for (iElem = 0; iElem < nElem; iElem++) {
 
-      node[iPoint]->SetElem(iElem);
+      /*--- Loop over all the nodes of an element ---*/
+      for (iNode = 0; iNode < elem[iElem]->GetnNodes(); iNode++) {
+        iPoint = elem[iElem]->GetNode(iNode);
+        elems[iPoint].push_back(iElem);
+      }
     }
+    nodes->SetElems(elems);
+  }
+  SU2_OMP_BARRIER
 
   /*--- Loop over all the points ---*/
 
-  for (iPoint = 0; iPoint < nPoint; iPoint++)
+  SU2_OMP_FOR_DYN(roundUpDiv(nPoint,2*omp_get_max_threads()))
+  for (iPoint = 0; iPoint < nPoint; iPoint++) {
 
-  /*--- Loop over all elements shared by the point ---*/
+    /*--- Loop over all elements shared by the point ---*/
 
-    for (iElem = 0; iElem < node[iPoint]->GetnElem(); iElem++) {
+    for (iElem = 0; iElem < nodes->GetnElem(iPoint); iElem++) {
 
-      jElem = node[iPoint]->GetElem(iElem);
+      jElem = nodes->GetElem(iPoint, iElem);
 
       /*--- If we find the point iPoint in the surronding element ---*/
 
-      for (iNode = 0; iNode < elem[jElem]->GetnNodes(); iNode++)
+      for (iNode = 0; iNode < elem[jElem]->GetnNodes(); iNode++) {
 
-        if (elem[jElem]->GetNode(iNode) == iPoint)
+        if (elem[jElem]->GetNode(iNode) != iPoint) continue;
 
         /*--- Localize the local index of the neighbor of iPoint in the element ---*/
 
-          for (iNeighbor = 0; iNeighbor < elem[jElem]->GetnNeighbor_Nodes(iNode); iNeighbor++) {
-            Node_Neighbor = elem[jElem]->GetNeighbor_Nodes(iNode, iNeighbor);
-            Point_Neighbor = elem[jElem]->GetNode(Node_Neighbor);
+        for (iNeighbor = 0; iNeighbor < elem[jElem]->GetnNeighbor_Nodes(iNode); iNeighbor++) {
+          Node_Neighbor = elem[jElem]->GetNeighbor_Nodes(iNode, iNeighbor);
+          Point_Neighbor = elem[jElem]->GetNode(Node_Neighbor);
 
-            /*--- Store the point into the point ---*/
-
-            node[iPoint]->SetPoint(Point_Neighbor);
-          }
+          /*--- Store the point into the point, if it is new ---*/
+          auto End = points[iPoint].end();
+          if (find(points[iPoint].begin(), End, Point_Neighbor) == End)
+            points[iPoint].push_back(Point_Neighbor);
+        }
+      }
     }
+    //sort(points[iPoint].begin(), points[iPoint].end());
 
-  /*--- Set the number of neighbors variable, this is
-   important for JST and multigrid in parallel ---*/
+    /*--- Set the number of neighbors variable, this is important for JST and multigrid in parallel. ---*/
+    nodes->SetnNeighbor(iPoint, points[iPoint].size());
+  }
 
-  for (iPoint = 0; iPoint < nPoint; iPoint++)
-    node[iPoint]->SetnNeighbor(node[iPoint]->GetnPoint());
+  SU2_OMP_MASTER
+  nodes->SetPoints(points);
```

Comment:
Significant change 1b: We set the adjacency for all points in one call, the input is a temporary vector<vector<long>> or similar which the function SetPoints compresses, this is only done in 3 places.